sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java

```diff
@@ -78,11 +81,13 @@ public String get(Object key) {
 
   @Override
   public String put(String key, String value) {
+    original.put(key, value);
```

Comment:
The thing worries me most is the inconsistency between the case insensitive map and the original map. I think we should either fail or keep the latter entry if `a -> 1, A -> 2` appears together.

One thing we can simplify is, `CaseInsensitiveStringMap` is read by data source and can be read-only. Then it can be easier to resolve conflicting entries at the beginning.

Reply:
But the method `put/remove/clear` still need to be implemented...Do you mean that we can ignore original map in these method ?

Reply:
We can just throw exception in these methods, and say this map is readonly.

Reply:
@cloud-fan This is a good solution! 
@rdblue What do you think?

Reply:
That works for me. It is always a good idea to avoid passing mutable state to plugin code.